v2.0.0: Implemented and moved all examples, documentation and tests over to namespaces data access attributes

AndriyMulyar opened the pull request:
# v2.0.0

Atlas map state is now accessible as top level attributes of the AtlasProjection ("the Map") class.
```python
from atlas import AtlasProject

project = AtlasProject(name='My Project')
map = project.maps[0]

```

- Topics (`map.topics`)
- Embeddings (`map.embeddings`)
- Semantic Duplicate Clusters (`map.duplicates`)

You can learn how to use the operations these attributes encapsulate on their corresponding data in the `Accessing Atlas State` section of the documentation.


------
Updated documentation and example scripts to reflect this change.

zanussbaum commented on docs/map_state/embeddings.md:
```diff
@@ -0,0 +1,24 @@
+
+map = project.maps[0]
+
+map.embeddings.df
```
My first natural question here is what columns can we expect? Will we get both the ambient and 2d representations in the same dataframe?

zanussbaum commented on docs/map_state/tags.md:
```diff
@@ -0,0 +1,26 @@
+
+map = project.maps[0]
+
+map.tags.get_tags()
```
Should we note here that tags returns a dictionary where the keys are the tag names the user adds in the frontend and the values are a list of ids? I found this confusing when I was initially using `get_tags`.

Should we also mention that if the use doesn't provide `id_field` in `map_text`, it's very tough to map from tags to your original dataframe? cc @bmschmidt

bmschmidt replied:
We should make sure all our examples show people uploading ids.

With the new auto assignment it's harder to screw it up... but still not great.

zanussbaum commented on docs/map_state/topics.md:
```diff
@@ -0,0 +1,25 @@
+
+map = project.maps[0]
+
+map.topics.df
```
nit: same thing here, can we either show an example or explain what the dataframe looks like (e.g. what columns)

AndriyMulyar replied:
Resolved by adding examples containing the dataframe structure

zanussbaum commented on nomic/atlas.py:
```diff
@@ -76,9 +78,7 @@ def map_embeddings(
-        data = [{
-            ATLAS_DEFAULT_ID_FIELD: b64int(i)
-        } for i in range(len(embeddings))]
+        data = [{ATLAS_DEFAULT_ID_FIELD: b64int(i)} for i in range(len(embeddings))]
```
it would be great if we either enforce users to upload with an id field or find a way to pass that to the user so they can map back from `datum_id` to their original data. I've had to rebuild maps a few too many times since I forgot to add an id column and couldn't join with the tags I had added to the data points

AndriyMulyar replied:
yeah, they just shouldn't be using map_embeddings or map_text.
its kinda a hard problem to solve.

AndriyMulyar replied:
I will add information that these are set in upload order.

bmschmidt replied:
I think we could write a warning to stderr on map_embeddings and map_text.

AndriyMulyar replied:
I will add a warning and also documentation.

zanussbaum replied:
What should people be using instead of map_embeddings or map_text then?

AndriyMulyar replied:
p = AtlasProject()

p.add_text()

p.create_index()

zanussbaum commented on nomic/data_operations.py:
```diff
@@ -1,42 +1,602 @@
+        """
+        return self.df
+
+    # @property
```
Should we update the documentation if this isn't finished yet?

zanussbaum commented on nomic/data_operations.py:
```diff
@@ -1,42 +1,602 @@
+        headers['Content-Type'] = 'application/octet-stream'
+        response = requests.post(self.project.atlas_api_path + "/v1/project/tag/delete", headers=headers, data=payload)
+        if response.status_code != 200:
+            raise Exception("Failed to delete tags")
```
do we want a `__repr__` method to stay consistent with the other classes?

AndriyMulyar replied:
What should this be. The issue with tags is that we don't get global access of all tag state

AndriyMulyar replied:
I will add a .df method for consistency

AndriyMulyar replied:
I have added a .df method which contains as columns a multi-hot vector where ones exist when the associated row has that tag.

zanussbaum replied:
yeah good point, resolving

zanussbaum commented on nomic/data_operations.py:
```diff
@@ -1,42 +1,602 @@
+
+    def download_embeddings(self, save_directory: str, num_workers: int = 10) -> bool:
+        '''
+        Downloads shards of arrow tables that map
```
nit: is this complete?

AndriyMulyar replied:
all of AtlasMapEmbeddings is not fully complete yet. it is blocked by ben implementing logic on the atlas cloud backend for writing embeddings more efficiently for export.

Once we have an efficient method pull out ambient embeddings, accessing with .ambient will be possible.

For now the existing logic where we download all embeddings to a directory is here.

zanussbaum commented on nomic/gpt4all/gpt4all.py:
```diff
@@ -23,19 +26,18 @@ def __init__(self, llama_path=None):
         self.model_path = llama_path
         self.tokenizer_path = llama_path
         self.lora_path = 'nomic-ai/vicuna-lora-multi-turn_epoch_2'
```
wait is the gpt4all model the first model we trained? should we update this to a newer one?

AndriyMulyar replied:
We should remove gpt4all from the repository, but lets do it in a separate PR

bmschmidt replied:
Removal should definitely be part of the 2.0 release.

AndriyMulyar replied:
🪓 done

zanussbaum reviewed (commented):
Had a few comments, mostly on documentation + clarifications

bmschmidt commented on docs/data_exploration_cleaning_tagging_in_atlas.ipynb:
```diff
@@ -172,7 +172,7 @@
    ],
    "source": [
-    "tags = map.get_tags()\n",
+    "tags = map.tags.get_tags()\n",
```
Ideally `map.tags` would do the retrieval on its own, not requiring a call to `map.tags.get_tags()`.

AndriyMulyar replied:
Done

rguo123 replied:
Notebook needs to be updated as well?

bmschmidt commented on docs/map_state/topics.md:
```diff
@@ -0,0 +1,25 @@
+Atlas pre-organizes your data into topics informed by the latent contents of your embeddings. Visually, these are represented by regions of homogenous
```
'regions of homogenous color' -> 'regions of the same color'

bmschmidt commented on docs/release_notes.md:
```diff
@@ -1,6 +1,24 @@
+- Semantic Duplicate Clusters (`map.duplicates`)
+- Tagging (`map.tags`)
+
+You can learn how to use the operations these attributes encapsulate on their corresponding data in the `Accessing Atlas State` section of the documentation.
```
Better--"The `Accessing Atlas State` section of the documentation shows some of the ways you can use this data."

bmschmidt commented on nomic/project.py:
```diff
@@ -8,46 +8,42 @@
-
 # Mainly for type checking
 try:
-    from pandas import DataFrame
     import pandas as pd
+    from pandas import DataFrame
 except ImportError:
     pd = None
     DataFrame = None
```
We can lose this try-catch; pandas is a dependency again.

bmschmidt commented on nomic/project.py:
```diff
@@ -366,14 +365,15 @@ def __init__(self, atlas_index_id, name, indexed_field, projections):
-    Manages operations on maps such as text/vector search.
+    Interact and access state of an Atlas Map including text/vector search.
     This class should not be instantiated directly.
     Instead instantiate an AtlasProject and use the project.indices or get_map method to retrieve an AtlasProjection.
```
Instead instantiate an AtlasProject and use the project.indices or get_map method to retrieve an AtlasProjection.
should be
    Instead instantiate an AtlasProject and use the project.maps attribute to retrieve an AtlasProjection.

bmschmidt commented on nomic/project.py:
```diff
@@ -1370,7 +1006,9 @@ def create_index(
-                    raise Exception(f"Could not find the index '{reuse_embeddings_from_index}' to re-use from. Possible options are {[index.name for index in indices]}")
+                    raise Exception(
+                        f"Could not find the index '{reuse_embeddings_from_index}' to re-use from. Possible options are {[index.name for index in indices]}"
+                    )
```
Seeing how aggressively we use the word 'Index' here, I'm wondering if we should make atlas.maps return all the indices, rather than the projections.... I think it will be confusing for people to encounter the idea of an index if we're calling them 'maps'

AndriyMulyar replied:
It is certainly confusing to have many abstraction, we have talked in person about this in-depth before.

Currently AtlasIndex is a proxy container from the ages where we were not certain what Atlas would be so built in room for flexibility.

I would propose we just *ignore* it and only mention AtlasProject and AtlasProjection ("map") across all documentation. We could essentially replace the word index to map in every location and no-one would be the wiser (and the code may be the clearer) as AtlasIndex is currently a passthrough object.

bmschmidt replied:
That sounds good to me--so this method will be renamed `reuse_embeddings_from_map`?

AndriyMulyar commented on docs/map_state/duplicates.md:
```diff
@@ -0,0 +1,26 @@
+Atlas groups your data into semantically similar clusters powered by latent information contained in your embeddings.
+Under the hood, it utilizes an algorithm similar to [SemDeDup](https://arxiv.org/abs/2303.09540).
+
+
+You can access and operate on semantic duplicate clusters programmatically by using the `duplicates` attribute
+of an AtlasMap. 
+
```
I think this would make a lot of sense.

AndriyMulyar commented on docs/map_state/embeddings.md:
```diff
@@ -0,0 +1,24 @@
+Atlas stores, manages and generates embeddings for your unstructured data.
+
+You can access Atlas embeddings in their ambient form (e.g. high dimensional) or in their two-dimensional
+projected representations.
```
embedded sounds confusing (map.embeddings.embedded?)

i am fine with either latent/ambient.


Another idea: what if we called them
two_dimensional
high_dimensional

instead?

bmschmidt replied:
I'm lukewarm on `two_dimensional`/`high_dimensional`. Partly because I don't like writing out numbers.

My thought on 'embedded' is that accessing embeddings does just by default work on the latent space, and the two-d space is a special case. It's not like 'embeddings.vector_search' lets you search in the 2d space as an option.

So maybe like `map.embeddings.array`, `map.embeddings.ndarray`, `map.embeddings.matrix` just returns the latent space as an `n x d` np.ndarray, and `map.embeddings.projected` returns the `n x 2` array.

AndriyMulyar commented:
- Removed references to GPT4All
- Added warning when people don't specify ids
- Gave `map.tags` a dataframe that can be sliced and diced (but it is materialized from the server state on every call)
- Added more code/output examples for all data state classes

zanussbaum commented on nomic/data_operations.py:
```diff
@@ -24,6 +24,27 @@
 class AtlasMapDuplicates:
     """
     Atlas Duplicate Clusters State
+
```
i know these were probably a pain to write up but this is very much appreciated and really improves on the docs 😄

AndriyMulyar replied:
No, this is what the world needed

zanussbaum commented on docs/map_state/embeddings.md:
```diff
@@ -7,18 +7,12 @@ projected representations.
 from nomic import AtlasProject
 
-project = AtlasProject(name='My Project')
+map = AtlasProject(name='My Project').maps[0]
```
Probably not the place for this, but most of my workflow has been creating a single map per project. What's the use case for multiple maps per single project?

AndriyMulyar replied:
if you index the project multiple times (say on different fields), you will have a map per index so possibly multiple indices

zanussbaum commented on nomic/atlas.py:
```diff
@@ -77,13 +77,19 @@ def map_embeddings(
             data[i][id_field] = b64int(i)
 
+    if added_id_field:
+        logger.warning("An ID field was not specified in your data so one was generated for you..")
```
Should we mention the ID is in order from `range(len(data))` in order that the data was added?

AndriyMulyar replied:
updated

zanussbaum reviewed (approved):
LGTM pending Ben's review

rguo123 commented on docs/map_state/duplicates.md:
````diff
@@ -0,0 +1,20 @@
+```python
+from nomic import AtlasProject
+
+map = AtlasProject(name='My Project').maps[0]
````
super nitpick: In one of the tutorials, it may be good to showcase `get_map` instead of accessing the list of maps?

rguo123 commented on docs/map_state/topics.md:
```diff
@@ -0,0 +1,20 @@
+Atlas pre-organizes your data into topics informed by the latent contents of your embeddings. Visually, these are represented by regions of homogenous
+color on an Atlas map. 
```
Nit: "AtlasMap" since that's used elsewhere

rguo123 commented on docs/vector_database.md:
```diff
@@ -6,7 +6,10 @@ This tutorial will show you how you can visualize your Weaviate and Pinecone vec
+Vector databases allow you to query your data semantically by indexing embedding vectors. By interactively visualizing embeddings, you can quickly:
+
+- Understand the space of possible query results from your vector database
+- Identify bad embeddings in your index which may produce poor query results.
```
Nit: no period

rguo123 commented on nomic/pl_callbacks/pl_callback.py:
```diff
@@ -45,8 +52,17 @@ def clear(self):
-    def __init__(self, max_points=-1, rebuild_time_delay=600, name=None, description=None, is_public=True, overwrite_on_validation=False):
+    def __init__(
+        self,
+        max_points=-1,
```
As part of docs, can we include what -1 is referring to here?

AndriyMulyar replied:
done
